driver-core/src/main/com/mongodb/connection/ByteBufPayloadBsonDocument.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.connection;
+
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonDocumentReader;
+import org.bson.BsonType;
+import org.bson.ByteBuf;
+
+import java.io.ByteArrayOutputStream;
+import java.io.InvalidObjectException;
+import java.io.ObjectInputStream;
+import java.io.Serializable;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+
+class ByteBufPayloadBsonDocument extends AbstractByteBufBsonDocument implements Cloneable, Serializable {
+    private static final long serialVersionUID = 3L;
+    private static final Charset UTF8_CHARSET = Charset.forName("UTF-8");
+
+    private final ByteBufBsonDocument commandDocument;
+    private final String payloadName;
+    private final List<ByteBufBsonDocument> payload;
+    private BsonDocument unwrapped;
+
+
+    static ByteBufPayloadBsonDocument create(final ByteBufferBsonOutput bsonOutput, final int startPosition) {
```

Comment:
I like it, but it's complicated enough that it could use some extract method refactoring so that it's clear what each step is doing.

Reply:
done.

Reply:
Much clearer this way.